src/gpuarray_reduction.c

```diff
@@ -256,113 +801,492 @@ static int   maxandargmaxCheckargs              (maxandargmax_ctx*  ctx){
 	ctx->nds = ctx->src->nd;
 	ctx->ndr = ctx->reduxLen;
 	ctx->ndd = ctx->nds - ctx->ndr;
+	strb_ensure(&ctx->s, 5*1024);
+
 
-	return ctx->ret;
+
+	return reduxSelectModel(ctx);
```

Comment:
In my previous version I tried to do all of the error handling in the highest-level function. That doesn't really scale as I add more and more and deeper functions, since each such function has to be able to detect the condition and return, and all its callers have to also detect a condition, abort and pass on the error until the highest level.

So I instead arranged the code in CPS style, and every function can abort the entire `GpuArray_reduction()` by tail-calling `reduxCleanup()` with whichever is the correct `GA_ERROR_CODE` as argument. A nice way of having clean "exception" handling in C.

CPS is somewhat unidiomatic C, but GCC does optimize my tailcalls from `callq` to `jmpq` in Release mode, or inlines them whole. And there is a net readability win from all the error handling code I've eliminated.